src/lib.rs

```diff
@@ -1222,6 +1222,7 @@ impl Async<UnixStream> {
         // The stream becomes writable when connected.
         stream.writable().await?;
 
+        stream.get_ref().peer_addr()?;
```

Comment:
```suggestion
        // On Linux, it appears the socket may become writable even when connecting fails, so we
        // must do an extra check here and see if the peer address is retrievable.
        stream.get_ref().peer_addr()?;
```

Reply:
I removed the Linux reference since I think you mentioned the same behaviour on macOS [here](https://github.com/stjepang/async-io/issues/18#issuecomment-681916927) :)

Reply:
nvm I just saw [your comment](https://github.com/stjepang/async-io/issues/18#issuecomment-682559099) :smile:

Reply:
Hmmm. It seems that was my mistake and I misspoke in the comment. On macOS, I get `ConnectionRefused` rather than `NotConnected`.

To make sure we're on the same page, this is my understanding:

When `UnixStream` is in non-blocking mode:

* On Linux, `Async::<UnixStream>::connect()` succeeds, but then the first following write errors with `NotConnected`.
* On macOS, `Async::<UnixStream>::connect()` fails immediately with `ConnectionRefused`.

When `UnixStream` is in blocking mode:

* On all systems, `UnixStream::connect()` blocks until the connection is really established.

The workaround here is specifically for Linux. In `Async::<UnixStream>::connect()`, we add an additional check at the end of the function that attempts to retrieve the `peer_addr()` so that we fail immediately rather than on the first following write.

Reply:
When I said "On macOS I get the NotConnected error in all cases." I meant to actually say:

On macOS I get the `ConnectionRefused` error with any of `blocking`, `async-std`, or `async-io`.

Reply:
I don't really know on which OS-es the `peer_addr()` check is necessary, but it probably doesn't hurt having it even on systems that don't need it.

Reply:
True, then let's just leave it like that for now I guess :)